lib/src/neu/generators/openapi3/openapi3-type-util.ts

```diff
@@ -0,0 +1,309 @@
+import assertNever from "assert-never";
+import {
+  areBooleanLiteralTypes,
+  areFloatLiteralTypes,
+  areIntLiteralTypes,
+  areStringLiteralTypes,
+  ArrayType,
+  dereferenceType,
+  isNotNullType,
+  isNullType,
+  isObjectType,
+  isReferenceType,
+  isStringLiteralType,
+  ObjectType,
+  ReferenceType,
+  Type,
+  TypeKind,
+  TypeTable,
+  UnionType
+} from "../../types";
+import {
+  ArraySchemaObject,
+  BooleanSchemaObject,
+  DiscriminatorObject,
+  IntegerSchemaObject,
+  NumberSchemaObject,
+  ObjectPropertiesSchemaObject,
+  ObjectSchemaObject,
+  ReferenceObject,
+  SchemaObject,
+  StringSchemaObject
+} from "./openapi3-schema";
+
+export function typeToSchemaOrReferenceObject(
+  type: Type,
+  typeTable: TypeTable,
+  nullable?: boolean
+): SchemaObject | ReferenceObject {
+  switch (type.kind) {
+    case TypeKind.NULL:
+      throw new Error("Null must be part of a union for OpenAPI 3");
+    case TypeKind.BOOLEAN:
+      return booleanSchema({ nullable });
+    case TypeKind.BOOLEAN_LITERAL:
+      return booleanSchema({ values: [type.value], nullable });
+    case TypeKind.STRING:
+      return stringSchema({ nullable });
+    case TypeKind.STRING_LITERAL:
+      return stringSchema({ values: [type.value], nullable });
+    case TypeKind.FLOAT:
+      return numberSchema({ format: "float", nullable });
+    case TypeKind.DOUBLE:
+      return numberSchema({ format: "double", nullable });
+    case TypeKind.FLOAT_LITERAL:
+      return numberSchema({ values: [type.value], format: "float", nullable });
+    case TypeKind.INT32:
+      return integerSchema({ format: "int32", nullable });
+    case TypeKind.INT64:
+      return integerSchema({ format: "int64", nullable });
+    case TypeKind.INT_LITERAL:
+      return integerSchema({ values: [type.value], format: "int32", nullable });
+    case TypeKind.DATE:
+      return stringSchema({ format: "date", nullable });
+    case TypeKind.DATE_TIME:
+      return stringSchema({ format: "date-time", nullable });
+    case TypeKind.OBJECT:
+      return objectTypeToSchema(type, typeTable, nullable);
+    case TypeKind.ARRAY:
+      return arrayTypeToSchema(type, typeTable, nullable);
+    case TypeKind.UNION:
+      return unionTypeToSchema(type, typeTable);
+    case TypeKind.REFERENCE:
+      return referenceTypeToSchema(type);
+    default:
+      assertNever(type);
+  }
+}
+
+function booleanSchema(
+  opts: {
+    values?: boolean[];
+    nullable?: boolean;
+  } = {}
+): BooleanSchemaObject {
+  const enums =
```

Comment:
This logic is duplicated, it could be abstracted in a separate function - e.g. : 

```
const enum = createEnum(options)
```

```
createEnum(options) {
 if (!options.values) {
   return;
 }

 if ( options.nullable) {
   return [...options.values, null];
 }

 return options.value;
} 
```